File: step_16_dynamic_programming/16.2_1d_dp/IB166_sub_matrices_with_sum_zero/solution.py
"""
Problem: Sub Matrices with Sum Zero
Difficulty: HARD | XP: 50
Source: InterviewBit

Given a 2D matrix, find the number of sub-matrices whose sum equals zero.
"""
from typing import List, Optional
from collections import defaultdict


# ============================================================
# APPROACH 1: BRUTE FORCE — Check all sub-matrices with prefix sum
# Time: O(m^2 * n^2)  |  Space: O(m * n)
# ============================================================
def brute_force(matrix: List[List[int]]) -> int:
    """
    Compute 2D prefix sums, then enumerate all possible sub-matrices
    defined by (r1, c1, r2, c2) and check if sum == 0.
    """
    if not matrix or not matrix[0]:
        return 0
    m, n = len(matrix), len(matrix[0])

    # Build prefix sum: prefix[i][j] = sum of matrix[0..i-1][0..j-1]
    prefix = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            prefix[i][j] = (matrix[i - 1][j - 1]
                            + prefix[i - 1][j]
                            + prefix[i][j - 1]
                            - prefix[i - 1][j - 1])

    count = 0
    for r1 in range(1, m + 1):
        for r2 in range(r1, m + 1):
            for c1 in range(1, n + 1):
                for c2 in range(c1, n + 1):
                    total = (prefix[r2][c2]
                             - prefix[r1 - 1][c2]
                             - prefix[r2][c1 - 1]
                             + prefix[r1 - 1][c1 - 1])
                    if total == 0:
                        count += 1
    return count


# ============================================================
# APPROACH 2: OPTIMAL — Fix row pair, use 1D subarray sum = 0
# Time: O(m^2 * n)  |  Space: O(n)
# ============================================================
def optimal(matrix: List[List[int]]) -> int:
    """
    Fix top and bottom rows. Compress columns into a 1D array (column sums).
    Then count subarrays with sum 0 using prefix sum + hashmap.
    """
    if not matrix or not matrix[0]:
        return 0
    m, n = len(matrix), len(matrix[0])
    count = 0

    for top in range(m):
        col_sum = [0] * n
        for bottom in range(top, m):
            for c in range(n):
                col_sum[c] += matrix[bottom][c]

            # Count subarrays of col_sum with sum 0
            prefix_count = defaultdict(int)
            prefix_count[0] = 1
            running = 0
            for c in range(n):
                running += col_sum[c]
                count += prefix_count[running]
                prefix_count[running] += 1

    return count


# ============================================================
# APPROACH 3: BEST — Same as optimal (this is the standard best approach)
# Time: O(m^2 * n)  |  Space: O(n)
# ============================================================
def best(matrix: List[List[int]]) -> int:
    """
    Same idea as optimal but swap m and n roles to use the smaller
    dimension for the outer loop, reducing constant factor.
    """
    if not matrix or not matrix[0]:
        return 0
    m, n = len(matrix), len(matrix[0])

    # Use the smaller dimension for the O(d^2) outer loop
    if m <= n:
        return _count_zero_submatrices(matrix, m, n)
    else:
        # Transpose
        transposed = [[matrix[i][j] for i in range(m)] for j in range(n)]
        return _count_zero_submatrices(transposed, n, m)


def _count_zero_submatrices(matrix, m, n):
    count = 0
    for top in range(m):
        col_sum = [0] * n
        for bottom in range(top, m):
            for c in range(n):
                col_sum[c] += matrix[bottom][c]

            prefix_count = defaultdict(int)
            prefix_count[0] = 1
            running = 0
            for c in range(n):
                running += col_sum[c]
                count += prefix_count[running]
                prefix_count[running] += 1
    return count


if __name__ == "__main__":
    print("=== Sub Matrices with Sum Zero ===")

    matrix = [[1, -1], [-1, 1]]
    print(f"Brute:   {brute_force(matrix)}")   # 5
    print(f"Optimal: {optimal(matrix)}")       # 5
    print(f"Best:    {best(matrix)}")          # 5

    matrix = [[0, 0], [0, 0]]
    print(f"Brute:   {brute_force(matrix)}")   # 9
    print(f"Optimal: {optimal(matrix)}")       # 9
    print(f"Best:    {best(matrix)}")          # 9

File: step_16_dynamic_programming/16.2_1d_dp/IB166_sub_matrices_with_sum_zero/Solution.java
/**
 * Problem: Sub Matrices with Sum Zero
 * Difficulty: HARD | XP: 50
 * Source: InterviewBit
 *
 * Given an N x N matrix, count all sub-matrices whose elements sum to 0.
 *
 * Approach: Fix two row boundaries r1 and r2. Compress the 2D problem
 * to 1D by summing columns between rows r1 and r2 (columnSum[j]).
 * Then count subarrays with sum 0 in columnSum[] using prefix sum + hashmap.
 *
 * Time: O(N^3)  -- O(N^2) row pairs * O(N) hashmap scan per pair.
 *
 * Real-life use: Financial analysis (zero-sum regions), image processing.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE
    // Time: O(N^5)  |  Space: O(1)
    // Enumerate all (r1, r2, c1, c2) combinations and sum each sub-matrix.
    // ============================================================
    public static int bruteForce(int[][] matrix) {
        int n = matrix.length;
        int count = 0;
        for (int r1 = 0; r1 < n; r1++) {
            for (int r2 = r1; r2 < n; r2++) {
                for (int c1 = 0; c1 < n; c1++) {
                    for (int c2 = c1; c2 < n; c2++) {
                        int sum = 0;
                        for (int r = r1; r <= r2; r++)
                            for (int c = c1; c <= c2; c++)
                                sum += matrix[r][c];
                        if (sum == 0) count++;
                    }
                }
            }
        }
        return count;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL
    // Time: O(N^3)  |  Space: O(N)
    // Fix r1 and r2, compute column prefix sums, then use hashmap
    // to count subarrays with sum 0 in O(N).
    // ============================================================
    public static int optimal(int[][] matrix) {
        int n = matrix.length;
        int count = 0;

        for (int r1 = 0; r1 < n; r1++) {
            int[] colSum = new int[n]; // colSum[j] = sum of matrix[r1..r2][j]
            for (int r2 = r1; r2 < n; r2++) {
                // Extend row range to r2
                for (int j = 0; j < n; j++) colSum[j] += matrix[r2][j];

                // Count subarrays with zero sum in colSum[]
                Map<Integer, Integer> prefixCount = new HashMap<>();
                prefixCount.put(0, 1);
                int prefixSum = 0;
                for (int j = 0; j < n; j++) {
                    prefixSum += colSum[j];
                    count += prefixCount.getOrDefault(prefixSum, 0);
                    prefixCount.merge(prefixSum, 1, Integer::sum);
                }
            }
        }
        return count;
    }

    // ============================================================
    // APPROACH 3: BEST
    // Time: O(N^3)  |  Space: O(N)
    // Same as optimal but with cleaner variable naming and uses
    // 2D prefix sums to pre-compute column sums in O(N^2) setup.
    // ============================================================
    public static int best(int[][] matrix) {
        int n = matrix.length;
        // Precompute 2D prefix sums for column range sums
        int[][] prefix = new int[n + 1][n]; // prefix[i][j] = sum of col j from row 0 to i-1
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                prefix[i + 1][j] = prefix[i][j] + matrix[i][j];

        int count = 0;
        for (int r1 = 0; r1 < n; r1++) {
            for (int r2 = r1; r2 < n; r2++) {
                // colSum[j] = sum of matrix[r1..r2][j]
                Map<Integer, Integer> freq = new HashMap<>();
                freq.put(0, 1);
                int ps = 0;
                for (int j = 0; j < n; j++) {
                    int colVal = prefix[r2 + 1][j] - prefix[r1][j];
                    ps += colVal;
                    count += freq.getOrDefault(ps, 0);
                    freq.merge(ps, 1, Integer::sum);
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        System.out.println("=== Sub Matrices with Sum Zero ===\n");

        int[][] m1 = {
            {1, -1},
            {-1,  1}
        };
        System.out.println("Test 1 (2x2 alternating, expected 5):");
        System.out.println("  Brute:   " + bruteForce(m1));
        System.out.println("  Optimal: " + optimal(m1));
        System.out.println("  Best:    " + best(m1));

        int[][] m2 = {
            { 1,  2, -3},
            {-1, -2,  3},
            { 0,  0,  0}
        };
        System.out.println("\nTest 2 (3x3):");
        System.out.println("  Brute:   " + bruteForce(m2));
        System.out.println("  Best:    " + best(m2));

        int[][] m3 = {{0}};
        System.out.println("\nTest 3 (single zero, expected 1):");
        System.out.println("  Best: " + best(m3));

        int[][] m4 = {
            {0, 0, 0},
            {0, 0, 0},
            {0, 0, 0}
        };
        System.out.println("\nTest 4 (all zeros 3x3, expected 36):");
        System.out.println("  Brute: " + bruteForce(m4));
        System.out.println("  Best:  " + best(m4));
    }
}
